### PR TITLE
Add a few more methods to FeatureList

### PR DESCRIPTION
**Pull request by cmyr**

### CODE REVIEW

**Comment by simoncozens** [comment]
This kind of annoying boilerplate is precisely my beef with the newtype pattern. It’s in almost every way a Vec, but to make it behave like a Vec we have to implement a bunch of trivially derivable methods on it.

Is it worth looking at something like Shrinkwrap, or is this just How It Is with newtypes?

**Comment by cmyr** [comment]
okay sorry, accidental book report incoming 🥴 

I actually think that maybe this doesn't really need to *be* a newtype, and that makes me feel like I should step back a bit and try to express when I think newtypes are useful, and when they are not.

## when do newtypes make sense

Basically: there are a few cases where I like newtypes:
- when I want a collection type that imposes more or different constraints on its members than some `std` type
-  when I really just want a typedef "with teeth"
- units & numeric types
- when I want to hide an implementation detail from public API

For the first case: let's say, or example, that I want to use a type to represent a sequence of codepoints. It is an invariant of my program that these codepoints are *always sorted*, and that there are no duplicates. If this type is just used internally in some private method, I would probably just use `Vec<u32>`, and I would ensure that if the collection is modified, it is sorted before it is used.

Now imagine that this collection is part of the public API, and the user needs to be able to add codepoints. If I just use `Vec<u32>`, it is trivial for them to modify the collection in such a way that it violates my invariants. In this case, I would almost certainly want to use a simple newtype, that exposed only the API that is needed, and which ensures the invariants are maintained, something like:

```rust
struct Codepoints(Vec<u32>);

impl Codepoints {
    /// returns `true` if this codepoint did not already exist in self.
    pub fn insert(&mut self, codepoint: u32) -> bool {
        match self.0.binary_search(&codepoint) {
            Ok(_) => false,
            Err(idx) => {
                self.0.insert(idx, codepoint);
                true
            }
        }
    }

   pub fn len(&self) -> bool {
        self.0.len()
    }
  // other methods as needed
}
```

For the second case, a relevant example is `GlyphId`. The OpenType spec uses `uint16` for all sorts of things, including class ids, glyph ids, point indices, various counts, etc. Here, newtypes can be helpful because they make it harder to write certain kinds of bugs, such as reversing the argument order in a method that accepts a glyph id as well as a contour point index. I *could* use a type alias for this, but in my experience they are rarely used consistently: it's easy to add a method that just takes `u16` where a `GlyphId` is intended. By making `GlyphId` an actual type, the compiler forces you to use it consistently.

## shrinkwrap and boilerplate

Sometimes this might be helpful, but sometimes it might be counterproductive. For instance in the `GlyphId` case, we may not want the user to `deref` to a `u16`, because our main goal is to treat the id as something distinct from a `u16`.

There is definitely always *some* boilerplate with these types, particular for methods like `len` and `is_empty`. This doesn't bother me too much personally, and it *personally* it bothers me less than having methods on the type (via, say, deref) that are not part of the API I'm trying to express: for instance if I have a `struct Stack<T>(Vec<T>);` I don't want the user to have access to things like the `get` method.

## this particular case

As written, the `FeatureList` type really isn't providing any advantage over the underlying `Vec`; the best evidence of this is the fact that the entire `impl` block is just calling methods on the inner collection. This *does* feel bad.

The bigger question for me, though, is *what are we trying to represent*? What are the constraints? For instance, is it valid for the same tag to appear in multiple entries in the vec? If not, this is the sort of thing that *might* justify a newtype, (although it might just suggest we should be using a `HashMap` instead).

## tl;dr

If we're going to use a newtype for a collection, it should be because we want to enforce some API constraint or otherwise better represent our application logic in a way that is not possible when using the collection directly.

## metapoint

If there had been a PR that introduced this newtype, I would have made the points I'm making here during code review. As is, though, I find it easier to just add the methods I need and which let me get my code working. If I were designing this API from scratch, there is certainly lots I would change, but I'm much less inclined to change API that already exists. I appreciate that you've been hacking on this a ton, and I'm grateful for all of the work: it would have probably taken me (at least!) the better part of a year to get up to speed and to get similar functionality finished, and I'm grateful for your commitment.

**Comment by cmyr** [comment]
My philosophy around a lot of this stuff (i.e. using the type system to enforce even fairly trivial program constraints) is: I *know* that I'm a dummy, and I *know* that I have in the past regularly overlooked trivial things, and every time I have an opportunity to make some sort of trivial mistake harder to make I am strongly inclined to consider it.

I do think there's real value in the "first make it work" approach. It is difficult to perfectly design an interface that you haven't used yet. This is one of the reasons that I'm really happy about the shift to a more layered design: it makes it much easier to replace the top layer (or experiment with variations on the top layer) without needing to rewrite everything else, and that will allow us to experiment much more easily with the API.

As an aside, I think this is actually a good rubric for deciding what should go in what layer: if someone were to write a replacement for just the `fonttools-rs` crate, all the code that they would need to call in order to generate the actual tables should be in `ot_spec`.
